### PR TITLE
Add --allarch and --arch to bodhi updates download

### PR DESCRIPTION
**Pull request by frantisekz**
This PR Adds support for following arguments in bodhi updates download:

--allarch - will download packages for all architectures
--arch [x86_64,i686,...] - will download package for specific architecture

### CODE REVIEW

**Comment by bowlofeggs** [comment]
jenkies test

**Comment by frantisekz** [comment]
@bowlofeggs Any idea what went wrong? I don´t see anything obvious in the test details :(

**Comment by bowlofeggs** [comment]
Hello @frantisekz!

Bodhi's tests use ```diff-cover``` to ensure that all altered lines in pull requests have test coverage:

```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged, and unstaged changes
-------------
bodhi/client/__init__.py (72.2%): Missing lines 460,462-463,500,502
-------------
Total:   18 lines
Missing: 5 lines
Coverage: 72%
-------------
```

There's a ```TestDownload``` class here that we can probably just drop a few tests in:

https://github.com/fedora-infra/bodhi/blob/develop/bodhi/tests/client/test___init__.py#L66

There's some good example tests that you can copy as a base, and I suggest just mocking the call to koji as is done in some of those, and asserting that the right call to koji happened.

Let me know if you would like more specific suggestions, I'd be happy to help!

**Comment by bowlofeggs** [comment]
By the way, if you have the Vagrant environment on your local setup, you can use the bash alias ```bdiff-cover``` in it to run this same test locally. It's much faster than Jenkins. It's also possible to run the tests in containers, but it'll be much easier to do that once https://github.com/fedora-infra/bodhi/pull/1870 is merged (it's just waiting for a review from somone in the fedora-infra group).

**Comment by frantisekz** [comment]
Hi @bowlofeggs ,
I've added tests for my additions. Unfortunately, it caused one other test to fail.
`_____________ ERROR collecting bodhi/tests/client/test___init__.py _____________
bodhi/tests/client/test___init__.py:372: in <module>
    class TestQuery(unittest.TestCase):
bodhi/tests/client/test___init__.py:407: in TestQuery
    return_value=client_test_data.EXAMPLE_QUERY_MUNCH_MULTI, autospec=True)
E   AttributeError: 'module' object has no attribute 'EXAMPLE_QUERY_MUNCH_MULTI'`

I can't figure out why is it happening. Do you have any idea?

Thanks!

File: bodhi/client/__init__.py
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: I think this approach is fine if you prefer it, but here are a couple other ideas:

0. Make it an error to specify both flags.
1. Have only the ```--arch``` flag and have it accept an "all" option.

I think I'd probably go for the second of those two, but if you really like the current approach that's fine with me.

File: bodhi/client/__init__.py
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: We could reduce the number of repeated koji snippets above a bit by doing something like this:

```
args = ['koji', 'download-build']
if all_arch is False
    args += ['--arch=noarch', '--arch={0}'.format(platform.machine())]
elif requested_arch is not None:
    args += ['--arch=noarch', '--arch={0}'.format(requested_arch)]
args += [build['nvr']]
```

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
I think this patch will get the tests passing:

```
diff --git a/bodhi/tests/client/test___init__.py b/bodhi/tests/client/test___init__.py
index ecdddfc..2c3b58c 100644
--- a/bodhi/tests/client/test___init__.py
+++ b/bodhi/tests/client/test___init__.py
@@ -107,7 +107,7 @@ class TestDownload(unittest.TestCase):
 
         result = runner.invoke(
             client.download,
-            ['--builds', 'nodejs-grunt-wrap-0.3.0-2.fc25', '--arch x86_64'])
+            ['--builds', 'nodejs-grunt-wrap-0.3.0-2.fc25', '--arch', 'x86_64'])
 
         self.assertEqual(result.exit_code, 0)
         self.assertEqual(result.output,
@@ -150,13 +150,16 @@ class TestDownload(unittest.TestCase):
 
         result = runner.invoke(
             client.download,
-            ['--builds', 'nodejs-grunt-wrap-0.3.0-2.fc25', '--allarch', '--arch x86_64'])
+            ['--builds', 'nodejs-grunt-wrap-0.3.0-2.fc25', '--allarch', '--arch', 'x86_64'])
 
         self.assertEqual(result.exit_code, 0)
-        self.assertEqual(result.output,
-                         'WARNING: --allarch suppressed by --arch\nDownloading packages from nodejs-grunt-wrap-0.3.0-2.fc25\n')
+        self.assertEqual(
+            result.output,
+            ('WARNING: --allarch suppressed by --arch\nDownloading packages from '
+             'nodejs-grunt-wrap-0.3.0-2.fc25\n'))
         call.assert_called_once_with((
-            'koji', 'download-build', '--arch=noarch', '--arch=x86_64', 'nodejs-grunt-wrap-0.3.0-2.fc25'))
+            'koji', 'download-build', '--arch=noarch', '--arch=x86_64',
+            'nodejs-grunt-wrap-0.3.0-2.fc25'))
 
     @mock.patch('bodhi.client.bindings.BodhiClient.csrf',
                 mock.MagicMock(return_value='a_csrf_token'))
```

The main thing was just making sure the --arch and x86_64 were separate parameters in the test arguments, and a couple long lines (which would fail flake8).

**Comment by frantisekz** [comment]
Thanks for feedback and explanation @bowlofeggs .
I´ve removed --allarch in favor of --arch all , seems more logical.

Regarding the tests, I´ve made changes you´ve suggested, but it´s still failing at the same point (I hopefully didn´t mess up anything else).

**Comment by bowlofeggs** [comment]
jenkies test

File: bodhi/client/__init__.py
Comment (bowlofeggs) [inline bodhi/client/__init__.py]: Let's tack on this at the end:

. "all" will retrieve packages from all architectures

File: bodhi/client/__init__.py
Comment (bowlofeggs) [inline bodhi/client/__init__.py:494]: Cool, this looks good!

File: bodhi/client/__init__.py
Comment (bowlofeggs) [inline bodhi/client/__init__.py:449]: s/./, or "all", indicating that all architectures are requested.

File: bodhi/tests/client/test___init__.py
Comment (bowlofeggs) [inline bodhi/tests/client/test___init__.py:138]: The tests look good to me, and they pass on my Vagrant box!

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
OK, this patch looks awesome! In addition to the two comments below, I have two more requests for you and we should be good to merge:

0. Can you add docs about this new flag to the man page? There's an existing section about downloads in there that has the other options documented, so we can just tack on a new bit about the new flag (and its "all" value): https://github.com/fedora-infra/bodhi/blob/develop/docs/man_pages/bodhi.rst
1. You will likely need to rebase onto develop to get the Jenkins to pass it, because I recently merged a PR that changed where CI test archive results get stored: https://github.com/fedora-infra/bodhi/pull/1873 - with a rebase you'll have the test code that puts the archive in the expected location and things should be peachy.

Thanks for the patch, nicely done!

**Comment by bowlofeggs** [comment]
I may have spoken too soon - all the CLI tests do pass, but I am seeing a test failure in one of the server tests: https://github.com/fedora-infra/bodhi/issues/1894

Is that what you were seeing? If so, it happens on the develop branch so it's definitely not your problem. I'll look into it.

**Comment by bowlofeggs** [comment]
Strangely, the server test that fails in vagrant actually does seem to pass in CI. In fact, it looks like both f25 and Rawhide passes with your current patch. So that's a bit weird, but I'll figure it out tomorrow.

The CI failure was just due to that moved file that I mentioned before, and you can just rebase to fix that.

**Comment by frantisekz** [comment]
I've rebased my PR on top of develop, but I am still seeing the same error in tests:

`_____________ ERROR collecting bodhi/tests/client/test___init__.py _____________
bodhi/tests/client/test___init__.py:351: in <module>
    class TestQuery(unittest.TestCase):
bodhi/tests/client/test___init__.py:386: in TestQuery
    return_value=client_test_data.EXAMPLE_QUERY_MUNCH_MULTI, autospec=True)
E   AttributeError: 'module' object has no attribute 'EXAMPLE_QUERY_MUNCH_MULTI'`

Looks like something different from #1894 though .

**Review by bowlofeggs — APPROVED** [review]
Looks good to me, I'll start the Jenkies tests and if they pass I'll squash your commits and merge.

As for the test failure, I'm not sure why you are seeing that. Are you able to confirm if ```bodhi/tests/client/__init__.py``` has a ```EXAMPLE_QUERY_MUNCH_MULTI``` varaible in it on your machine? I don't see that problem with your patch locally, and CI also doesn't seem to see it.

Thanks for the patch!

**Comment by bowlofeggs** [comment]
jenkies test

**Comment by bowlofeggs** [comment]
I'm 99.9% sure this is going to fail again, due to https://github.com/fedora-infra/bodhi/pull/1903 - once again - it's not your fault. Once that is merged, we'll need to rebase one more time. I'm really sorry for so much churn - I promise that it's not typical for Bodhi ☺ I recently switched to these container based CI tests and I think some bugs are still shaking out. It'll be more solid soon.

I'll see if GitHub will let me do the rebase for you - possibly no action needed on your part.

**Comment by bowlofeggs** [comment]
jenkies test

**Comment by frantisekz** [comment]
Okay, the issue was that I was running tests on top of bodhi 2.12.2 tag and not on develop. It passes the tests fine right now. Thanks for feedback help with the PR!

**Comment by bowlofeggs** [comment]
Ah, glad that was explained! I'm merging this, thanks for the patch!

**Pull request merged**
